examples/output-layout.c

```diff
@@ -124,94 +141,161 @@ static void handle_output_frame(struct output_state *output,
+	struct sample_keyboard *keyboard = wl_container_of(listener, keyboard, key);
+	struct sample_state *sample = keyboard->sample;
+	struct wlr_event_keyboard_key *event = data;
+	uint32_t keycode = event->keycode + 8;
+	const xkb_keysym_t *syms;
+	int nsyms = xkb_state_key_get_syms(keyboard->device->keyboard->xkb_state,
+			keycode, &syms);
+	for (int i = 0; i < nsyms; i++) {
+		xkb_keysym_t sym = syms[i];
+		if (sym == XKB_KEY_Escape) {
+				wl_display_terminate(sample->display);
+			}
+		// NOTE: It may be better to simply refer to our key state during each frame
+		// and make this change in pixels/sec^2
+		// Also, key repeat
+		int delta = 75;
+		if (event->state == WLR_KEY_PRESSED) {
+			switch (sym) {
+			case XKB_KEY_Left:
+				update_velocities(sample, -delta, 0);
+				break;
+			case XKB_KEY_Right:
+				update_velocities(sample, delta, 0);
+				break;
+			case XKB_KEY_Up:
+				update_velocities(sample, 0, -delta);
+				break;
+			case XKB_KEY_Down:
+				update_velocities(sample, 0, delta);
+				break;
+			}
+		}
+	}
 }
 
-static void update_velocities(struct compositor_state *state,
-		float x_diff, float y_diff) {
-	struct sample_state *sample = state->data;
-	sample->x_vel += x_diff;
-	sample->y_vel += y_diff;
+void keyboard_destroy_notify(struct wl_listener *listener, void *data) {
+	struct sample_keyboard *keyboard = wl_container_of(listener, keyboard, destroy);
+	wl_list_remove(&keyboard->destroy.link);
+	wl_list_remove(&keyboard->key.link);
+	free(keyboard);
 }
 
-static void handle_keyboard_key(struct keyboard_state *kbstate,
-		uint32_t keycode, xkb_keysym_t sym, enum wlr_key_state key_state,
-		uint64_t time_usec) {
-	// NOTE: It may be better to simply refer to our key state during each frame
-	// and make this change in pixels/sec^2
-	// Also, key repeat
-	int delta = 75;
-	if (key_state == WLR_KEY_PRESSED) {
-		switch (sym) {
-		case XKB_KEY_Left:
-			update_velocities(kbstate->compositor, -delta, 0);
-			break;
-		case XKB_KEY_Right:
-			update_velocities(kbstate->compositor, delta, 0);
-			break;
-		case XKB_KEY_Up:
-			update_velocities(kbstate->compositor, 0, -delta);
-			break;
-		case XKB_KEY_Down:
-			update_velocities(kbstate->compositor, 0, delta);
-			break;
+void new_input_notify(struct wl_listener *listener, void *data) {
+	struct wlr_input_device *device = data;
+	struct sample_state *sample = wl_container_of(listener, sample, new_input);
+	struct sample_keyboard *keyboard;
+	struct xkb_rule_names rules;
+	struct xkb_context *context;
+	switch (device->type) {
+	case WLR_INPUT_DEVICE_KEYBOARD:
+		keyboard = calloc(1, sizeof(struct sample_keyboard));
+		keyboard->device = device;
+		keyboard->sample = sample;
+		wl_signal_add(&device->events.destroy, &keyboard->destroy);
+		keyboard->destroy.notify = keyboard_destroy_notify;
+		wl_signal_add(&device->keyboard->events.key, &keyboard->key);
+		keyboard->key.notify = keyboard_key_notify;
+		memset(&rules, 0, sizeof(rules));
```

Comment:
Use a designated initialiser instead of doing this. You could also make the scope a lot tighter.

Reply:
This was hoisted directly from the old `support/shared.c` FTR, but yeah I'll fix that.

Reply:
Ah correction no it was not. For some reason I thought C behaved like C++ in the behaviour of switch statements. Fixing now.